packages/data/src/types.d.ts

```diff
@@ -0,0 +1,36 @@
+export type WPDataFunctionOrGeneratorArray = {
+    [index: number]: Function|Generator;
+};
+export type WPDataFunctionArray = {
+    [index: number]: Function;
+};
+
+export type WPDataAttachedStore = {
+    getSelectors: () => WPDataFunctionArray,
+    getActions: () => WPDataFunctionArray,
+    subscribe: (listener: () => void) => (() => void)
+};
+
+export type WPDataStore = {
+    /**
+     * Store Name
+     */
+    name: string,
+
+    /**
+     * Store configuration object.
+     */
+    instantiate: (registry: WPDataRegistry) => WPDataAttachedStore,
+};
+
+export type WPDataReduxStoreConfig = {
+    reducer: ( state: any, action: any ) => any,
+    actions?: WPDataFunctionOrGeneratorArray,
+    resolvers?: WPDataFunctionOrGeneratorArray,
+    selectors?: WPDataFunctionArray,
+    controls?: WPDataFunctionArray,
+}
```

Comment:
It's generally preferred to use `interface` over `type`, I'd suggest rewriting all of these in the following way:

```suggestion
export interface WPDataReduxStoreConfig {
    reducer: ( state: any, action: any ) => any,
    actions?: WPDataFunctionOrGeneratorArray,
    resolvers?: WPDataFunctionOrGeneratorArray,
    selectors?: WPDataFunctionArray,
    controls?: WPDataFunctionArray,
}
```

Reply:
>It's generally preferred to use interface over type, I'd suggest rewriting all of these in the following way:

Is there a place where we could learn when to use one or the other?

Reply:
So here's my take on it after reading some things on line:

 - Use interface for public APIs when you can.
 - Use type where interface usage is not possible.

but it seems in general, typescript is making them more and more similar over time. Am I wrong?

Reply:
[This article](https://ncjamieson.com/prefer-interfaces/) seems to reflect the current state of things, or specifically [this tweet](https://twitter.com/drosenwasser/status/1319205169918144513) from the TypeScript program manager:

> Honestly, my take is that it should really just be interfaces for anything that they can model. There is no benefit to type aliases when there are so many issues around display/perf.

It has to do with a lot of quirks I don't fully understand myself 🙂 

> but it seems in general, typescript is making them more and more similar over time. Am I wrong?

They're somehow fundamentally different in the type system and I'm not aware of them becoming more similar, although it [does sound like something they may consider](https://twitter.com/drosenwasser/status/1319205566393192448).